### ISSUE TITLE
[FYI] - Newer BlackVue cameras have a rudimentary firewall protecting web server

### ISSUE THREAD

**Issue by kquinsland** [acolomba/blackvuesync#17]
**I am posting this as a PSA/FYI** as I spent a few hours banging my head against this problem.
This is not an issue with the code in this repo; I am posting in the hopes that the next guy doesn't waste time on the stuff I already solved :D.

------

I hit this issue with the `DR970X_PLUS` camera but I'd assume that it applies to other models in the same generation and possibly older generations with recent firmware.

## Issue

All requests - from a different subnet - are silently dropped before reaching the web server on the camera.
For the sake of simplicity/demonstration, I'll use the `192.168.10/24` and `192.168.90/24` as the two subnets; the camera lives on `.90` and the sync tool on `.10`.

```shell
# From a host in 192.168.10 space running the sync tool
# Note that the dashcam lives on the .90 network, not the .10
❯ curl -vvv http://dashcam.my-lan.internal/  
* Host dashcam.my-lan.internal:80 was resolved.  
* IPv6: (none)  
* IPv4: 192.168.90.24  
*   Trying 192.168.90.24:80…  
* connect to 192.168.90.24 port 80 from 192.168.10.15 port 39390 failed: Connection  
timed out  
* Failed to connect to dashcam.my-lan.internal port 80 after 134465 ms: Co  
uldn't connect to server  
* Closing connection  
curl: (28) Failed to connect to dashcam.my-lan.internal port 80 after 1344  
65 ms: Couldn't connect to server
```

But when I do the same request from another host on the `.90` network, it works fine.

```shell
# From a host in the 192.168.90 space; same subnet as the dashcam
❯ curl -vvv http://dashcam.my-lan.internal/  
* Host dashcam.my-lan.internal:80 was resolved.  
* IPv6: (none)  
* IPv4: 192.168.90.24  
*   Trying 192.168.90.24:80...  
* Connected to dashcam.my-lan.internal (192.168.90.24) port 80  
> GET / HTTP/1.1  
> Host: dashcam.my-lan.internal  
> User-Agent: curl/8.8.0  
> Accept: */*  
>    
* Request completely sent off  
< HTTP/1.1 200 OK  
< Date: Wed, 31 Jul 2024 02:03:57 GMT  
< Server: DR970X_PLUS  
< Accept-Ranges: bytes  
< Connection: Keep-Alive  
< Keep-Alive: timeout=10, max=1000  
< Content-Length: 59  
< Last-Modified: Tue, 09 Jul 2024 06:40:53 GMT  
< Content-Type: text/html  
<    
<html>  
<body>  
<H1>BlackVue</H1><br>  
  
</body>  
</html>  
* Connection #0 to host dashcam.my-lan.internal left intact
```

## Cause

So there's a basic filter on the camera that looks at the IP address of the incoming request and silently drops it if it's not from the same network segment as the camera (`.90` network in this example).

## Solution

The solution is to do source NAT/SNAT on whatever host(s) are responsible for routing between the subnets.
Effectively, you need a rule that says:

>  if the destination is `$dashcam` and the `$sourceIP` is not in the `.90` network, rewrite the `$sourceIP` to one that lives on the `.90` network

This way, the camera will always see requests as coming from another device on the same network and will happily respond to them. The device that did the SNAT will then re-write the packet and send it down to the originating requester on the other subnet.

Doing so is an exercise left to the reader as it depends on the routing setup in your network...etc.

**Comment by tfindlay-au** [comment]
FWIW - If you are trying to secure your camera, it seems they have Google's DNS hard coded in where it wants to call home to `pitta.blackviewcloud.com`

```
11:24:12.743975 IP 10.1.30.8.49695 > 8.8.8.8.53: 19004+ A? pitta.blackvuecloud.com. (41)
11:24:12.746874 IP 10.1.30.8.59650 > 8.8.4.4.53: 19004+ A? pitta.blackvuecloud.com. (41)
```

**Comment by kquinsland** [comment]
I am sorry to necro an old thread and send out notifications.

I did a semi-regular check in on this script and noticed that it hadn't been able to connect to the camera and pull clips for a while. It took some digging but I can confirm that there's some "new" behavior in play, at least with the new `2.00x` firmware series.

TL;DR: the camera will disconnect from the wireless network ~10s after connecting to WiFi if it can't set up a https connection to: 

- `gateway.blackvuecloud.com`
- `pitta.blackvuecloud.com` 

Shortly after the camera associates with the AP and does the DHCP dance, it will attempt DNS resolution of `gateway` and do https request where it will then DNS for `pitta`. If either fails, it will disconnect from the AP and will not attempt to connect to the wireless network again until you pull the power to reboot the camera. A device that isn't on your LAN is a device that you can't pull footage from!

The audible "unregistered camera" notification is the simplest "yep, it's going to stay connected to the AP" signal. You will not get this unless TCP/443 is permitted to the IP(s) that the hostnames above resolve to.

I don't have the time to pop open the firmware and figure out what could be done to change this behavior nor do I have the time to see if cert pinning is in effect / use my DNS server to send traffic to my own https server/MITM device.

-----

It took a while to narrow this down but the protocol I used to test and confirm:

- Create a "from: $dashCam, allow anything to WAN" rule. Ensure that is is in effect.
- Create a "from: $dashCam, deny anything except NTP/DNS/DHCP" rule. Ensure it is disabled.
- Power camera up, monitor your wireless network logs to confirm the MAC associated with your camera is able to connect
- Observe the usual "unregistered camera" voice message (assuming you have not actually bothered to register yours...)
- Swap the enable/disable on the firewall rules so the "only dhcp/ntp/dns is OK" rule is in effect.
- Reboot your camera or have the AP kick it off.
- Monitor wireless network logs for the same MAC again. You'll observe a "connected" event and then no more than 15s later you'll see a "disconnected" event.
- You will not hear the "unregistered camera" notification.
- revert the firewall rule enable/disable and reboot the camera and it'll stay connected to the LAN, now.


You can "simulate" driving by disabling the AP and toggling your ignition from standby to on/run. Wait 60s and turn the AP back on
As long as the "deny everything except the essentials" rule is in place, you'll see the same connect -- 15s elapses --> disconnect cycle.

Again swap the firewall rules around and then perform the same driving simulation and you'll notice the device stays connected.
